### PR TITLE
Add parts_of_speech rules, update syntax rules, and improve checks.

### LINKED ISSUE #18: Lookup problem with "correct"
image

### LINKED ISSUE #19: Lookup problem with "saw"
image

### LINKED ISSUE #25: lookup problem with "law-B/Law"
image

image

maybe the lookup needs to be a little smarter about case sensitivity... the Ontology app may also need to be capable of case-sensitive searches.

### PR DESCRIPTION
**Pull request by craigp-atw**
- Rework syntax check to use built-in rules.
- Add checks to disambiguate parts of speech.
- Improve handling of pairings. Instead of a Pairing token, it is a Lookup token with a complex pairing. Then it can be treated like any other lookup token for filtering and checking.
- Improve handling of pronoun referents. Make it a lookup token with a pronoun. Then it can be treated like any other lookup token for filtering and checking.
- Allow a Lookup token to show an error message. The resulting error token still has the ontology link.
- Fix some edge cases.
- Add tests.

This covered a bit more than I originally planned, but it ultimately fixes several issues (#18, #19, #25) and paves the way for more (#30, #31)

**Fixes #18:**
`John was correct. John was correct/just. John will judge/correct.`
- `John was correct.` - correct is still ambiguous here between the Verb or the Adjective (for now, until the right rule is added later).
- `John was correct/just.` - correct the Adjective (simple) is chosen since it matches 'just'.
- `John will judge/correct.` - correct the Verb (complex) is chosen since it matches 'judge'. Judge the Verb was chosen over the Noun because it matches 'correct'.
image

**Fixes #19:**
`John saw a saw.` - the parts of speech are determined by rules. `John saw X.` no longer shows as complex.
image

**Fixes #25:**
`John kept the law/Law. John kept the Law/scripture.`
- `John kept the law/Law.` - Law now shows as complex
- `John kept the Law/scripture.` - Law is now determined to be complex and so results in a level error.
image

`The Law is good. Law said, ["Law."]`
- `The Law is good.` - Law now shows as complex
- `Law said, ["Law."]` - Law remains ambiguous at the start of a sentence and quote.
image

**Pairings with different parts of speech results in an error:**
`John was happy/disciple`
image

**Differentiate uses of 'that':**
`The person [that guards the house]. The person sees that guard.`
- `The person [that guards the house].` - 'that' here is a relativizer or complementizer, not the demonstrative. So 'guard' cannot (yet) be disambiguated between Noun and Verb.
- `The person sees that guard.` - 'that' here is the demonstrative, so guard is determined to be the (complex) Noun.
image

**Readded 'speak' with quote check to prove usefulness of tagging quote clauses:**
`John spoke to Mary, ["Yes."]` - also shows link to ontology on the error token
image

### CODE REVIEW

File: app/src/lib/lookups.js [hunk 1]
```diff
@@ -68,11 +70,6 @@ async function check_forms(lookup_token) {
 	const results = await response.json()
 
 	lookup_token.form_results = results.matches
-
```
Comment (craigp-atw) [inline app/src/lib/lookups.js:71]: Moved this check to syntax_rules.json

File: app/src/lib/lookups.js [hunk 2]
```diff
@@ -107,7 +104,7 @@ async function check_ontology(lookup_token) {
 	if (lookup_token.form_results.length) {
 		const forms = lookup_token.form_results
 		lookup_token.lookup_results = lookup_token.lookup_results
-			.filter(lookup => forms.some(form => form.part_of_speech === lookup.part_of_speech))
+			.filter(lookup => forms.some(form => form.form === 'stem' || form.part_of_speech === lookup.part_of_speech))
```
Comment (craigp-atw) [inline app/src/lib/lookups.js:107]: Fixes for a word like 'correct' where the form lookup just results in the Adjective, but the ontology lookup results in both the Adjective and Verb. Since 'correct' is an uninflected form, it cannot be considered exclusive to the one part of speech.

File: app/src/lib/parser/pairings.test.js [hunk 3]
```diff
@@ -47,9 +47,18 @@ function create_lookup_result(stem, {sense='A', part_of_speech='Noun', level=1}=
...
+function create_sentence(tokens) {
+	return { clause: create_clause_token(tokens) }
+}
+
 describe('pairing level check', () => {
 	test('both words right level', () => {
-		const test_tokens = [
+		const test_tokens = [create_sentence([
 			create_pairing_token(
 				create_lookup_token('first', {lookup_results: [create_lookup_result('first', {level: 0})]}),
```
Comment (craigp-atw) [inline app/src/lib/parser/pairings.test.js:63]: At some point it might be good to rework the tests to use a mock function for lookups. I could add an issue for that?

File: app/src/lib/rules/rules_processor.js [hunk 4]
```diff
@@ -16,7 +14,7 @@ export function rules_applier(rules) {
  * @returns {Sentence[]}
  */
 export function apply_rules(sentences, rules) {
-	return sentences.map(sentence => create_sentence(apply_rules_to_tokens(sentence.clause.sub_tokens, rules)))
+	return apply_rules_to_tokens(sentences.map(sentence => sentence.clause), rules).map(clause => ({ clause }))
```
Comment (craigp-atw) [inline app/src/lib/rules/rules_processor.js:17]: Needed rules to apply to the main clause token as well.

File: app/src/lib/rules/checker_rules.js
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/lib/rules/checker_rules.js:1]: How would you feel about renaming these files to exclude `rules`?  It seems redundant to me.  I'd imagine the following if you agreed it was a little more concise:

```bash

rules/checker.js
rules/part_of_speech.js
rules/parser.js
rules/processor.js
rules/syntax.js

```

File: app/src/lib/rules/syntax_rules.js
Comment (longrunningprocess) [inline app/src/lib/rules/syntax_rules.js:1]: more AWESOME❣️

**Review by longrunningprocess — DISMISSED** [review]
This PR is written so well 🤩 It could easily serve as documentation and a reference for future developers by answering so many of "why?" questions, thank you for taking the time to make this level of documentation so important.

**Comment by craigp-atw** [comment]
Just discovered that the lookup filtering by case has a bug that affects something like `you(Paul)`. Pronoun referents somehow need to be handled differently. This should get resolved before merging.

File: app/src/lib/lookups.js
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/lookups.js:1]: Moved to a regex (like you noticed) and the case check is now in syntax_rules.js

File: app/src/lib/parser/index.js [hunk 5]
```diff
@@ -15,18 +15,19 @@ export async function parse(text) {
...
 		clausify,
-		check_capitalization,
 	)(text)
 
 	const with_forms = await perform_form_lookups(pre_lookups)
 	const with_transformed_lookups = rules_applier(LOOKUP_RULES)(with_forms)
 	const with_lookups = await perform_ontology_lookups(with_transformed_lookups)
 
 	return pipe(
+		rules_applier(SYNTAX_RULES),
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/parser/index.js:25]: Yes, there's no way around a sequence constraint. The order is quite important with the rules. We can discuss more about it if you want.

File: app/src/lib/parser/token.js [hunk 6]
```diff
@@ -122,7 +120,9 @@ export function token_has_concept(token) {
  */
 export function token_has_error(token) {
 	return token.type === TOKEN_TYPE.ERROR
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/parser/token.js:122]: Yes, I have an idea for the next change/PR that we can chat about.

File: app/src/lib/rules/checker_rules.js
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/rules/checker_rules.js:1]: I still like it with `_rules` to be honest, it makes it clear that the file only contains rules, especially with the tabs open in VSCode. Once migrated to store the rules in the db, we won't need those files anymore at all.

File: app/src/lib/rules/part_of_speech_rules.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import {parse_part_of_speech_rule} from './rules_parser'
+
+/**
+ * TODO store these in the database
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/rules/part_of_speech_rules.js:4]: Good, idea, I'll do that and remove the TODOs

File: app/src/lib/rules/checker_rules.js
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/lib/rules/checker_rules.js:1]: sounds good

File: app/src/lib/parser/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/lib/parser/index.js:25]: no need... it happened to me too so I figured it may be inevitable

File: app/src/lib/parser/token.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/lib/parser/token.js:124]: looks like we have a plan for `error_message` driving these error determinations going forward.

**Review by longrunningprocess — APPROVED** [review]
paired through remaining discussions, good to go.
